src/model/operation/transform.js

```diff
@@ -1,644 +1,1895 @@
+		operation.baseVersion = baseVersion++;
+	}
+
+	return operations;
+}
+
+function transform( a, b, context = { isStrong: false } ) {
+	const transformationFunction = getTransformation( a, b );
+
+	return transformationFunction( a.clone(), b, context );
+}
+
+function transformSets( operationsA, operationsB, options ) {
+	operationsA = operationsA.slice();
+	operationsB = operationsB.slice();
+
+	if ( operationsA.length == 0 || operationsB.length == 0 ) {
+		return { operationsA, operationsB };
+	}
+
+	const data = {
+		nextBaseVersionA: operationsA[ operationsA.length - 1 ].baseVersion + 1,
+		nextBaseVersionB: operationsB[ operationsB.length - 1 ].baseVersion + 1,
+		extraOpsA: 0,
+		extraOpsB: 0
+	};
+
+	const context = {};
+
+	if ( options.useContext ) {
+		initializeContext( context, operationsA, operationsB, options.document );
+	}
+
+	for ( let i = 0; i < operationsA.length; ) {
+		const opsA = [ operationsA[ i ] ];
+
+		for ( let j = 0; j < operationsB.length; ) {
+			const opsB = [ operationsB[ j ] ];
+
+			for ( let k = 0; k < opsA.length; ) {
+				for ( let l = 0; l < opsB.length; ) {
+					const opA = opsA[ k ];
+					const opB = opsB[ l ];
+
+					const { contextAB, contextBA } = getCurrentContext( context, opA, opB );
+
+					const newOpA = transform( opA, opB, contextAB );
+					const newOpB = transform( opB, opA, contextBA );
+
+					if ( options.useContext ) {
+						updateOriginalOperations( context, opA, newOpA );
+						updateOriginalOperations( context, opB, newOpB );
+
+						updateRelations( context, opA, opB );
+					}
+
+					opsA.splice( k, 1, ...newOpA );
+					k = k + newOpA.length;
+
+					opsB.splice( l, 1, ...newOpB );
+					l = l + newOpB.length;
+				}
+			}
+
+			operationsB.splice( j, 1, ...opsB );
+			j = j + opsB.length;
+			data.extraOpsB += opsB.length - 1;
+		}
+
+		operationsA.splice( i, 1, ...opsA );
+		i = i + opsA.length;
+		data.extraOpsA += opsA.length - 1;
+	}
+
+	if ( options.padWithNoOps ) {
+		padWithNoOps( operationsA, data.extraOpsB - data.extraOpsA );
+		padWithNoOps( operationsB, data.extraOpsA - data.extraOpsB );
+	}
+
+	updateBaseVersions( operationsA, data.nextBaseVersionB );
+	updateBaseVersions( operationsB, data.nextBaseVersionA );
+
+	return { operationsA, operationsB };
+}
+
+export default {
+	transform,
+	transformSets
+};
+
+function padWithNoOps( operations, howMany ) {
+	for ( let i = 0; i < howMany; i++ ) {
+		operations.push( new NoOperation( 0 ) );
+	}
+}
+
+function getCurrentContext( context, opA, opB ) {
+	if ( !context.document ) {
+		return {
+			contextAB: { isStrong: true },
+			contextBA: { isStrong: false }
+		};
+	}
+
+	const contextAB = {
+		isStrong: true,
+		aWasUndone: context.wasUndone( opA ),
+		bWasUndone: context.wasUndone( opB ),
+		relation: context.getRelation( opA, opB )
+	};
+
+	const contextBA = {
+		isStrong: false,
+		aWasUndone: contextAB.bWasUndone,
+		bWasUndone: contextAB.aWasUndone,
+		relation: context.getRelation( opB, opA )
```

Comment:
All of these could be calculated on demand.

Reply:
If `isStrong` will have to be a function, it will be bad in case of `transform` function:

```javascript
function transform( a, b, context = { isStrong: false } ) {
	const transformationFunction = getTransformation( a, b );

	return transformationFunction( a.clone(), b, context );
}
```

And transformation units tests, which we will still need some.